SubscribeToken never completes if WebSocket gets closed

I was recently upgrading our service dependencies and upgraded this module from `v.1.1.1` -> `v1.3.5` and found that one of our integration tests started failing. The scenario essentially creates a client and connects to a broker. It then attempts to subscribe to a topic that does not exist. This results in the following client error being received. `[ERROR] [client]   Connect comms goroutine - error triggered websocket: close 1005 (no status)`. This error is read from the `ibound` channel after the out bound priority `packets.SubscribePacket` is written to the connection.

The main issue here is that the call to `token.Wait()` **blocks forever**.  I tried to dig through and understand what's happening and I think the issue is that when this error gets read from the `ibound` channel ([net.go](https://github.com/eclipse/paho.mqtt.golang/blob/3f4740799ae33c4a62f3a6f6a818e019f1188860/net.go#L192)) is gets passed to another channel and eventually read off of the `commsErrors` channel ([client.go](https://github.com/eclipse/paho.mqtt.golang/blob/3f4740799ae33c4a62f3a6f6a818e019f1188860/client.go#L637)) but never gets propagated to the original SubscribeToken, thus meaning the `baseToken.complete` channel is never populated or closed.

I found that if I propagated the connection error by adding `c.index[c.lastIssuedID].setError(err)` below `ERROR.Println(CLI, "Connect comms goroutine - error triggered", err)`, the call to `token.Wait()` did return. I doubt this is a proper fix, but perhaps help in some way?

The broker used in our integration testing is an AWS IoT Core broker, and the mqtt client is configured like so:

```go
connOpts := iotMQTT.NewClientOptions()
connOpts.AddBroker(url)
connOpts.SetClientID(clientId)
connOpts.SetCleanSession(true)
connOpts.SetAutoReconnect(true)
connOpts.SetKeepAlive(2 * time.Second)

iotClient := iotMQTT.NewClient(connOpts)
token := iotClient.Connect()
if !token.Wait() || token.Error() != nil {
	return nil, fmt.Errorf("error connecting to iot: %s", token.Error())
}

token := iotClient.Subscribe("non-existing-topic", 1, nil)
token.Wait() // <- blocks forever when error "websocket: close 1005 (no status)" is received.
```
I found that if I upgraded from `v1.1.1` -> `v1.2.0` I encounter the same issue, but with a different error message. I believe various changes since `v1.2.0` have changed the error message/handling, but I only dug through `v1.3.5` as we'd like to use the latest version.

I've attached the stdout logs with debug enabled.
[stdout_log.txt](https://github.com/eclipse/paho.mqtt.golang/files/7460395/stdout_log.txt)

The AWS Broker logs the following message when attempting to subscribe to the `non-existing-topic`:
```json
{
    "timestamp": "2021-11-02 12:15:42.720",
    "logLevel": "ERROR",
    "traceId": "OMITTED",
    "accountId": "OMITTED",
    "status": "Failure",
    "eventType": "Subscribe",
    "protocol": "MQTT",
    "topicName": "non-existing-topic",
    "clientId": "OMITTED",
    "principalId": "OMITTED",
    "sourceIp": "OMITTED",
    "sourcePort": 59414,
    "reason": "AUTHORIZATION_FAILURE",
    "details": "Authorization Failure"
}
```

--- a/client.go
+++ b/client.go
@@ -514,7 +514,9 @@ func (c *client) internalConnLost(err error) {
 			reconnect := c.options.AutoReconnect && c.connectionStatus() > connecting
 
 			if c.options.CleanSession && !reconnect {
-				c.messageIds.cleanUp()
+				c.messageIds.cleanUp() // completes PUB/SUB/UNSUB tokens
+			} else if !c.options.ResumeSubs {
+				c.messageIds.cleanUpSubscribe() // completes SUB/UNSUB tokens
 			}
 			if reconnect {
 				c.setConnected(reconnecting)
@@ -811,7 +813,9 @@ func (c *client) Subscribe(topic string, qos byte, callback MessageHandler) Toke
 	}
 	DEBUG.Println(CLI, sub.String())
 
-	persistOutbound(c.persist, sub)
+	if c.options.ResumeSubs { // Only persist if we need this to resume subs after a disconnection
+		persistOutbound(c.persist, sub)
+	}
 	switch c.connectionStatus() {
 	case connecting:
 		DEBUG.Println(CLI, "storing subscribe message (connecting), topic:", topic)
@@ -883,7 +887,9 @@ func (c *client) SubscribeMultiple(filters map[string]byte, callback MessageHand
 		sub.MessageID = mID
 		token.messageID = mID
 	}
-	persistOutbound(c.persist, sub)
+	if c.options.ResumeSubs { // Only persist if we need this to resume subs after a disconnection
+		persistOutbound(c.persist, sub)
+	}
 	switch c.connectionStatus() {
 	case connecting:
 		DEBUG.Println(CLI, "storing subscribe message (connecting), topics:", sub.Topics)
@@ -1096,7 +1102,9 @@ func (c *client) Unsubscribe(topics ...string) Token {
 		token.messageID = mID
 	}
 
-	persistOutbound(c.persist, unsub)
+	if c.options.ResumeSubs { // Only persist if we need this to resume subs after a disconnection
+		persistOutbound(c.persist, unsub)
+	}
 
 	switch c.connectionStatus() {
 	case connecting:

--- a/messageids.go
+++ b/messageids.go
@@ -42,6 +42,7 @@ const (
 	midMax uint16 = 65535
 )
 
+// cleanup clears the message ID map; it sets an error and completes PUB, SUB and UNSUB tokens.
 func (mids *messageIds) cleanUp() {
 	mids.Lock()
 	for _, token := range mids.index {
@@ -52,6 +53,11 @@ func (mids *messageIds) cleanUp() {
 			token.setError(fmt.Errorf("connection lost before Subscribe completed"))
 		case *UnsubscribeToken:
 			token.setError(fmt.Errorf("connection lost before Unsubscribe completed"))
+		case *DisconnectToken:
+			token.setError(fmt.Errorf("connection lost before Disconnect completed"))
+		case *ConnectToken:
+			// Connect token will be closed in the connect function
+			continue
 		case nil:
 			continue
 		}
@@ -60,6 +66,27 @@ func (mids *messageIds) cleanUp() {
 	mids.index = make(map[uint16]tokenCompletor)
 	mids.Unlock()
 	DEBUG.Println(MID, "cleaned up")
+}
+
+// cleanUpSubscribe removes all SUBSCRIBE and UNSUBSCRIBE tokens (setting error)
+// This may be called when the connection is lost and we will not be resending SUB/UNSUB packets
+func (mids *messageIds) cleanUpSubscribe() {
+	mids.Lock()
+	for mid, token := range mids.index {
+		switch token.(type) {
+		case *SubscribeToken:
+			token.setError(fmt.Errorf("connection lost before Subscribe completed"))
+			delete(mids.index, mid)
+		case *UnsubscribeToken:
+			token.setError(fmt.Errorf("connection lost before Unsubscribe completed"))
+			delete(mids.index, mid)
+		case nil:
+			continue
+		}
+		token.flowComplete()
+	}
+	mids.Unlock()
+	DEBUG.Println(MID, "cleaned up subs")
 }
 
 func (mids *messageIds) freeID(id uint16) {

--- a/messageids.go
+++ b/messageids.go
@@ -42,7 +42,7 @@ const (
 	midMax uint16 = 65535
 )
 
-// cleanup clears the message ID map; it sets an error and completes PUB, SUB and UNSUB tokens.
+// cleanup clears the message ID map; completes all token types and sets error on PUB, SUB and UNSUB tokens.
 func (mids *messageIds) cleanUp() {
 	mids.Lock()
 	for _, token := range mids.index {
@@ -53,12 +53,7 @@ func (mids *messageIds) cleanUp() {
 			token.setError(fmt.Errorf("connection lost before Subscribe completed"))
 		case *UnsubscribeToken:
 			token.setError(fmt.Errorf("connection lost before Unsubscribe completed"))
-		case *DisconnectToken:
-			token.setError(fmt.Errorf("connection lost before Disconnect completed"))
-		case *ConnectToken:
-			// Connect token will be closed in the connect function
-			continue
-		case nil:
+		case nil: // should not be any nil entries
 			continue
 		}
 		token.flowComplete()
@@ -69,7 +64,7 @@ func (mids *messageIds) cleanUp() {
 }
 
 // cleanUpSubscribe removes all SUBSCRIBE and UNSUBSCRIBE tokens (setting error)
-// This may be called when the connection is lost and we will not be resending SUB/UNSUB packets
+// This may be called when the connection is lost, and we will not be resending SUB/UNSUB packets
 func (mids *messageIds) cleanUpSubscribe() {
 	mids.Lock()
 	for mid, token := range mids.index {
@@ -80,10 +75,7 @@ func (mids *messageIds) cleanUpSubscribe() {
 		case *UnsubscribeToken:
 			token.setError(fmt.Errorf("connection lost before Unsubscribe completed"))
 			delete(mids.index, mid)
-		case nil:
-			continue
 		}
-		token.flowComplete()
 	}
 	mids.Unlock()
 	DEBUG.Println(MID, "cleaned up subs")

